V1.4-Update use case in DG (#250)
AY2324S2-CS2103-F09-2/tp: docs/DeveloperGuide.md

### 6.3 Use cases

(For all use cases below, the **System** is `PatientSync` and the **Actor** is the `Nurse`, unless specified otherwise)

**Use case: UC01 - Add a patient**

**MSS**

1. Nurse requests to add a patient.
1. PatientSync adds the patient and display success message.

    Use case ends.

**Extensions**

* 1a. Nurse's input is invalid command format.

    * 1a1. PatientSync shows an error message.
           
      Use case ends.
  
* 1b. Nurse's input contains invalid parameters value. 

>     * 1b1. PatientSync shows an error message. 

Reviewer: Perhaps can add one more use case where Nurse added patient with duplicated patient hospital ID?

Author: do you think it can be counted as invalid parameters value?

Reviewer: ya I think should be can!

    
      Use case ends.

**Use case: UC02 - Delete a patient**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.
1. Nurse requests to delete a specific patient in the list.
1. PatientSync deletes the patient and displays success message.

    Use case ends.

**Extensions**

* 2a. The list is empty.
      Use case ends.

* 3a. Nurse's input is invalid command format.

    * 3a1. PatientSync shows an error message.
    
      Use case ends.

* 3b. Nurse's input contains invalid patient id.

    * 3b1. PatientSync shows an error message.
    
      Use case ends.

**Use case: UC03 - Edit a patient**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.

> 1. Nurse requests to edit a specific patient in the list.

Reviewer: Would it be better to phrase it as `Nurse requests to edit a specific patient's information/ field in the list`?

Author: good suggestion! changed

> 1. PatientSync edit the patient and displays success message.

Reviewer: Would it be better to phrase it as` PatientSync edits the patient information and displays success message` ?

Author: good suggestion! changed

   Use case ends.

**Extensions**

* 2a. The list is empty.
  Use case ends.

> * 3a. Nurse's input is invalid command format.

Reviewer: Would be better if it is phrased to `Nurse's input command format is invalid`?

  * 3a1. PatientSync shows an error message.

    Use case ends.

> * 3b. Nurse's input contains invalid patient id.

Reviewer: Would suggest to change -> Nurse's input contains invalid ~~patient id~~ `patientHospitalId` / `patient's hospital ID` (depending on the other use cases format)

Author: the patient ID here should be referring to the id in the patient list, thus not `patientHospitalId`

Reviewer: Would it be better if we use index instead?

  * 3b1. PatientSync shows an error message.

    Use case ends.

* 3c. Nurse's input contains invalid parameters value.

  * 3c1. PatientSync shows an error message.

>     Use case ends.

Reviewer: Would be great as well if we could add one more use case where the index to edit inputted by nurse is invalid? or do you think it is considered under the invalid command format use case?

Author: I think it can be considered under invalid parameters!

### 6.3 Use cases

(For all use cases below, the **System** is `PatientSync` and the **Actor** is the `Nurse`, unless specified otherwise)

**Use case: UC01 - Add a patient**

**MSS**

1. Nurse requests to add a patient.
1. PatientSync adds the patient and display success message.

    Use case ends.

**Extensions**

* 1a. Nurse's input is invalid command format.

    * 1a1. PatientSync shows an error message.
           
      Use case ends.
  
* 1b. Nurse's input contains invalid parameters value.

    * 1b1. PatientSync shows an error message. 
    
      Use case ends.

* 1b. Nurse's input contains duplicate patient hospital Id.

  * 1b1. PatientSync shows an error message.

    Use case ends.

**Use case: UC02 - Delete a patient**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.
1. Nurse requests to delete a specific patient in the list.
1. PatientSync deletes the patient and displays success message.

    Use case ends.

**Extensions**

* 2a. The list is empty.
      Use case ends.

* 3a. Nurse's input is invalid command format.

    * 3a1. PatientSync shows an error message.
    
      Use case ends.

> * 3b. Nurse's input contains invalid patient id.

Reviewer: should it be `ID`/ `Id` instead if `id`?

Author: Good catch!

    * 3b1. PatientSync shows an error message.
    
      Use case ends.

**Use case: UC03 - Edit a patient's information**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.
1. Nurse requests to edit a specific patient's information in the list.
1. PatientSync edits the patient information and displays success message.

   Use case ends.

**Extensions**

* 2a. The list is empty.
  Use case ends.

* 3a. Nurse's input is invalid command format.

  * 3a1. PatientSync shows an error message.

    Use case ends.

> * 3b. Nurse's input contains invalid patient id.

Reviewer: Similar to the comment above regarding `id`.

Author: Good catch!

  Use case ends.

* 3a. Nurse's input is invalid command format.

    * 3a1. PatientSync shows an error message.

      Use case resumes at step 2.

* 3b. Nurse's input contains invalid patient index or event index.

    * 3b1. PatientSync shows an error message.

      Use case resumes at step 2.

**Use case: UC07 - Edit an event for a patient**

**Preconditions: Event exists for the target patient**

**MSS**

1.  Nurse requests to list patients.
1.  PatientSync shows a list of patients.
1.  Nurse requests to edit an event for a specific patient in the list.

> 1.  PatientSync edits an event the patient and displays success message.

Reviewer 2: ```suggestion
1.  PatientSync edits an event for the patient and displays success message.
```

Author: Good catch! updated!

  Use case ends.

* 3a. Nurse's input is invalid command format.

  * 3a1. PatientSync shows an error message.

    Use case ends.

* 3b. Nurse's input contains invalid patient index or event index.

  * 3b1. PatientSync shows an error message.

    Use case ends.

**Use case: UC08 - Add tag to a patient**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.
1. Nurse requests to add a tag to a specific patient in the list.

> 1. PatientSync add a tag to the patient and displays success message.

Reviewer: Would be good if `add` change to `adds`

Author: good catch!

  Use case ends.

* 3a. Nurse's input is invalid command format.

  * 3a1. PatientSync shows an error message.

    Use case ends.

* 3b. Nurse's input contains invalid patient index or event index.

  * 3b1. PatientSync shows an error message.

    Use case ends.

> 

Reviewer 2: ```suggestion
* 3c. Nurse's input contains invalid event name.
    * 3c1. PatientSync shows an error message.
    Use case resumes at step 2.
* 3d. Nurse's input contains invalid date or datetime format.
    * 3d1. PatientSync shows an error message.
    Use case resumes at step 2.
* 3e. Nurse's input contains duplicate event.
    * 3e1. PatientSync shows an error message.
    Use case resumes at step 2.
```
These are the other cases where there will be an error message.

Author: Updated!

  Use case ends.

* 3a. Nurse's input is invalid command format.

  * 3a1. PatientSync shows an error message.

    Use case resumes at step 2.

* 3b. Nurse's input contains invalid patient index or event index.

  * 3b1. PatientSync shows an error message.

    Use case resumes at step 2.

**Use case: UC08 - Add tag to a patient**

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.

> 1. Nurse requests to add a tag to a specific patient in the list.

Reviewer 3: one or more tags

Author: Updated !

> 1. PatientSync adds a tag to the patient and displays success message.

Reviewer 3: adds the specified tag(s) and display status message

Author: Updated!

  Use case ends.

* 3a. Nurse's input is invalid command format.

    * 3a1. PatientSync shows an error message.

      Use case resumes at step 2.

* 3b. Nurse's input contains invalid patient index.

  * 3b1. PatientSync shows an error message.

    Use case resumes at step 2.

**Use case: UC09 - Delete tag from a patient**

> **Preconditions: Tag exists for the target patient**

Reviewer 3: better not to use preconditions in use case and put it under extensions

Author: Good point! removed the precondition!

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.

> 1. Nurse requests to delete a tag from a specific patient in the list.

Reviewer 3: one or more tags

Author: Updated!

> 1. PatientSync deletes a tag from the patient and displays success message.

Reviewer 3: delete the specified tag(s) in the patient's tag list

Author: Updated!

  Use case ends.

* 3a. Nurse's input is invalid command format.

  * 3a1. PatientSync shows an error message.

    Use case resumes at step 2.

* 3b. Nurse's input contains invalid patient index.

  * 3b1. PatientSync shows an error message.

    Use case resumes at step 2.

* 3c. Nurse's input contains invalid tags.

  * 3c1. PatientSync shows an error message.

    Use case resumes at step 2.

> 

Reviewer 3: * 3d. The patient selected does not contain specified tag
* 3d1. PatientSync shows an error message.
Use case resumes at step 2

Author: Updated!

> **Use case: UC10 - Find patients by a tag**

Reviewer 2: ```suggestion
**Use case: UC10 - Find patients by tag**
```

Author: updated to tag(s)! to be consistent with below comments!

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.

> 1. Nurse requests to find patients with a specific tag in the list.
> 1. PatientSync finds patients with the specified tag and displays success message.

Reviewer 2: ```suggestion
1. Nurse requests to find patients with specific tag(s) in the list.
1. PatientSync finds patients with the specified tag(s) and displays success message.
```

Reviewer 2: User can find using 1 or more tags

Author: Nice catch! updated!

    Use case ends.

**Extensions**

* 2a. The list is empty.

  Use case ends.

* 3a. Nurse's input is invalid command format.

    * 3a1. PatientSync shows an error message.

      Use case resumes at step 2.

> **Use case: UC11 - Find patients by keyword**

Reviewer 2: ```suggestion
**Use case: UC11 - Find patients by name**
```

Author: It is being consistent with the Section on locating patients by name 3.10. Maybe should standardise both?

Author: After offline discussion, it is ok to change to name!

**MSS**

1. Nurse requests to list patients.
1. PatientSync shows a list of patients.

> 1. Nurse requests to find patients with specific keywords in the list.
> 1. PatientSync finds patients with the specified keywords and displays success message.

Reviewer 2: ```suggestion
1. Nurse requests to find patients with specific name(s) in the list.
1. PatientSync finds patients with the specified name(s) and displays success message.
```

Author: It is being consistent with the Section on locating patients by name 3.10. Maybe should standardise both?

Author: After offline discussion, it is ok to change to name!
